allium/lib/Parser/Parser.cpp

```diff
@@ -44,56 +45,56 @@ Optional<PredicateDecl> Parser::parsePredicateDecl() {
         Parameter param;
 
         do {
-            if(parseParameter().unwrapInto(param)) {
+            if(parseParameter().unwrapResultInto(param, errors)) {
                 parameters.push_back(param);
             } else {
                 if (parameters.size() == 0) {
-                    emitSyntaxError("Parentheses must not appear after predicate name for predicates with zero arguments.");
+                    errors.push_back(SyntaxError("Parentheses must not appear after predicate name for predicates with zero arguments.", lexer.peek_next().location));
                 } else {
-                    emitSyntaxError("Expected an additional argument after \",\" in parameter list.");
+                    errors.push_back(SyntaxError("Expected an additional parameter after \",\" in parameter list.", lexer.peek_next().location));
                 }
-
-                return rewindAndReturn();
             }
         } while(lexer.take(Token::Type::comma));
 
         if(!lexer.take(Token::Type::paren_r)) {
-            emitSyntaxError("Expected a \",\" or \")\" after parameter.");
-            return rewindAndReturn();
+            errors.push_back(SyntaxError("Expected a \",\" or \")\" after parameter.", lexer.peek_next().location));
         }
 
         std::vector<EffectRef> effects;
-        if(parseEffectList().unwrapGuard(effects)) {
+        if(parseEffectList().unwrapResultGuard(effects, errors)) {
             rewindAndReturn();
```

Comment:
This probably should have been `return rewindAndReturn()`... but now it should just be `lexer.rewind(identifier)`.

Reply:
Done in ee82e67b9e5228d233df48110b6018895b4c1df7. Will remove if needed based on discussion on the other `rewindAndReturn()` comment.